ext/DifferentiationInterfaceTapirExt/allocating.jl

```diff
@@ -0,0 +1,26 @@
+function DI.value_and_pullback(f, ::AutoTapir, x, dy, rrule)
+    y = f(x)  # TODO: one call too many, just for the conversion
+    dy_righttype = convert(typeof(y), dy)
+    _, (_, dx) = value_and_pullback!!(rrule, dy_righttype, f, x)
+    return y, dx
+end
+
+function DI.value_and_pullback!!(f, dx, ::AutoTapir, x, dy, rrule)
+    y = f(x)  # TODO: one call too many, just for the conversion
+    dy_righttype = convert(typeof(y), dy)
+    dx_righttype = zero_sametype!!(dx, x)
+    new_y, (_, new_dx) = value_and_pullback!!(
+        rrule, dy_righttype, zero_codual(f), CoDual(x, dx_righttype)
```

Comment:
for some reason it errors if I replace `zero_codual(f)` with `f` here

Reply:
```julia
MethodError: Cannot `convert` an object of type Tapir.Tangent{@NamedTuple{x::Float64, dx::Float64}} to an object of type Tapir.Stack{Float64}
```
what does this mean?

Reply:
(note that this is not related to my issues with the mutating aspect)

Reply:
Yup, you'll need to make every argument a `CoDual`, or none of them. This is a really uninformative error message though, so I'll make an issue on my end to improve it.

Reply:
right, thanks